src/util.cc

```diff
@@ -1652,14 +1684,47 @@ int choose_cpu(BindCPU bind_cpu) {
   // presumably due to cheaper context switching and/or
   // better interaction with CPU frequency scaling.
   if (bind_cpu >= 0) {
+    if (cpu_lock_fd_out.is_open()) {
+      struct flock lock {
+        .l_type = F_WRLCK,
+        .l_whence = SEEK_SET,
+        .l_start = bind_cpu,
+        .l_len = 1,
+        .l_pid = 0
+      };
+      // Try to acquire the lock for this CPU
+      (void)fcntl(cpu_lock_fd_out, F_SETLK, &lock);
+      // Ignore fcntl errors - nothing we can do
+    }
     return bind_cpu;
   }
 
-  int cpu = get_random_cpu_cgroup();
-  if (cpu >= 0) {
-    return cpu;
+  if (cpu_lock_fd_out.is_open()) {
+    // Try twice to allocate a CPU. If we fail twice, pick a random one
+    for (int i = 0; i < 2; ++i) {
+      std::shuffle (cpus.begin(), cpus.end(), std::default_random_engine(random()));
+      for (int cpu : cpus) {
+        struct flock lock {
+          .l_type = F_WRLCK,
+          .l_whence = SEEK_SET,
+          .l_start = cpu,
+          .l_len = 1,
+          .l_pid = 0
+        };
+        // Try to acquire the lock for this CPU
+        int err = fcntl(cpu_lock_fd_out, F_SETLK, &lock);
+        if (err == 0)
+          return cpu;
+        else if (err == -1) {
+          DEBUG_ASSERT(errno == EACCES || errno == EAGAIN);
+        }
+      }
+    }
   }
-  return random() % get_num_cpus();
+
+  // Didn't work - just use a random CPU
+  cpu_lock_fd_out.close();
```

Comment:
I don't think you need `close()` since ScopedFd will clean up?

Reply:
Yes, it will - at the end of the session. That's probably fine, since in this situation we didn't acquire any locks, but I thought cleaning up explicitly would be nicer, in case we ever need to look at it for debugging or something. That way we'd know if it failed to acquire the lock.

Reply:
Ah right. No, this is fine.

BTW can you confirm that we don't leak the lock fd into the tracee?

Reply:
It's cloexec'ed, so it should be gone after the initial exec.